core/src/avm2/optimizer/optimize.rs

```diff
@@ -1944,3 +1926,171 @@ fn abstract_interpret_ops<'gc>(
     )?;
     Ok(())
 }
+
+fn optimize_get_property<'gc>(
+    activation: &mut Activation<'_, 'gc>,
+    multiname: Gc<'gc, Multiname<'gc>>,
+    stack_value: OptValue<'gc>,
+) -> Result<Option<(Op<'gc>, Option<Class<'gc>>)>, Error<'gc>> {
+    if let Some(vtable) = stack_value.vtable() {
+        match vtable.get_trait(&multiname) {
+            Some(Property::Slot { slot_id }) | Some(Property::ConstSlot { slot_id }) => {
+                let mut value_class = vtable.slot_classes()[slot_id as usize];
+                let resolved_value_class = value_class.get_class(activation)?;
+
+                vtable.set_slot_class(activation.gc(), slot_id as usize, value_class);
+
+                return Ok(Some((Op::GetSlot { index: slot_id }, resolved_value_class)));
+            }
+            Some(Property::Virtual {
+                get: Some(disp_id), ..
+            }) => {
+                let full_method = vtable
+                    .get_full_method(disp_id)
+                    .expect("Method should exist");
+                let ClassBoundMethod { method, .. } = full_method;
+                if !method.is_info_resolved() {
+                    method.resolve_info(activation)?;
+                }
+
+                let declared_params = method.resolved_param_config();
+                let return_type = method.resolved_return_type();
+
+                let mut result_op = Op::CallMethod {
+                    num_args: 0,
+                    index: disp_id,
+                    push_return_value: true,
+                };
+
+                // We can further optimize calling FastCall getters into a
+                // static native method call
+                if let MethodKind::Native {
```

Comment:
Any chance this pattern
```rs
if let MethodKind::Native {
    native_method,
    fast_call: true,
} = method.method_kind()
{
    if stack_value.class.is_some_and(|c| c.is_final()) {
        if /* check len(arguments)==N and types match */
        {
            result_op = Op::CallNative {
                method: *native_method,
                num_args: 1,
                push_return_value: false,
            };
        }
    }
```
Can be refactored to a function or macro? I see 4 variations of it here, just differing by # of args.